### ISSUE TITLE
ILQL consumes more GPU memory than PPO

### ISSUE THREAD

**Issue by hzwer** [CarperAI/trlx#390]
I tried to deploy PPO and ILQL algorithms with the same bloom3B model under examples/summarize_rlhf/, and changed the reward model to a naive calculation. My GPU is A100 with 32GB.
I need to adjust the batchsize and sequence length to be smaller to run ILQL, and deploying PPO is very smooth.
I don't understand why ILQL needs more GPU memory.
Any suggestions?

**Comment by hzwer** [comment]
In the process of dealing with GPU memory overhead, I found that reducing the deepspeed version to 0.7.5 is really useful
https://github.com/microsoft/DeepSpeed/issues/3076

**Comment by reciprocated** [comment]
Hi! ILQL does a bit more allocations than PPO, but it also uses more/and bigger heads on top of the base LLM, so that's one reason for more GPU memory usage. Are you also using the same `num_unfrozen_layers` for both setups? Because in most configs in examples here, PPO freezes all but the last 2 layers of the base LLM, while for ILQL all parameters are usually left as trainable, which should also be of a major contribution to GPU memory usage.

About the deepspeed issue, it's really interesting. How much memory were you able to save after downgrading? Also the original issue only mentions about the difference in the context of inference and doesn't say anything about training
🤔

**Comment by hzwer** [comment]
I set the same frozen layers and use lora finetuning. I will check the head later.
As for deepspeed, we found 10% memory saving both for RL and sft. I mention it here for people that find methods to deal with GPU memory issues.

**Comment by hzwer** [comment]
I get it. ILQL's Q head is very memory-intensive and time-consuming, so it can be scaled down appropriately
